### ISSUE TITLE
Space group symmetry

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#1437]
It could be useful to support simulations with explicit space group symmetry, along the lines of http://pubs.acs.org/doi/abs/10.1021/ct100506d.  I think this could be done with just another type of virtual site, one that calculates its position by applying a symmetry transform to a single other particle.

**Comment by peastman** [comment]
I'm looking into how to implement this.  I *think* the following design should work.  To define the virtual site, you will specify the index of another particle, a 3x3 matrix **R**, and three coefficients u, v, and w.  The virtual site's position is then set to

**Rx** + u**a** + v**b** + w**c**

where **x** is the position of the other particle, and **a**, **b**, and **c** are the current box vectors.  Will that cover all the intended use cases?

**Comment by jchodera** [comment]
How are the space group symmetries (e.g. P212121) encoded here?

**Comment by peastman** [comment]
I'm not sure what you mean.  They're encoded by a rotation and a translation.

**Comment by jchodera** [comment]
I think there's a bit of a disconnect between what we envision for the API here.

From a user perspective, I want to be able to invoke a simulation by specifying a desired [crystallographic space group](https://en.wikipedia.org/wiki/List_of_space_groups) along with the fact that I want PME.  The PDB standard uses the [Hermann-Mauguin space group symbol](http://deposit.rcsb.org/adit/docs/pdb_atom_format.html#CRYST1) using the full symbol, such as `P 21 21 2`.  There are some [statistics on space group frequencies in the PDB here](http://www.rcsb.org/pdb/statistics/histogram.do?mdcat=symmetry&mditem=space_group_name_H-M&numOfbars=200&name=Space%20Group).  FFX goes further to ensure that the unit cell parameters are also consistent with the internal symmetry of the space group.

As a user, I would hope for a simple way to specify which of the 230 space groups I want through this same terminology, which opens us the possibility for simulating crystals from the PDB directly.  Something like:
```python
system = ForceField.createSystem(topology, nonbondedMethod=PME, spacegroup='P 21 21 2', ...)
```
If there is a lower-level C++ API that utilizes a virtual site scheme, that's fine, but we should think of it from the high-level user-facing perspective too.

**Comment by peastman** [comment]
My comments above strictly related to the lower level API for describing the calculations.  We can also discuss a high level modeling API, but that isn't what I was referring to.

**Comment by jchodera** [comment]
> My comments above strictly related to the lower level API for describing the calculations. We can also discuss a high level modeling API, but that isn't what I was referring to.

How about we worth from both ends? I think there's value in ensuring that we present the user with something they can actually use; you need to define a concrete low-level API to actually implement---ideally one that makes the high-level implementation straightforward.

**Comment by peastman** [comment]
Sounds good.

**Comment by peastman** [comment]
I've been thinking about how to do a high level API for this, and there are a couple of interesting complications.

First of all, when you start adding symmetric copies, that suddenly means that global translations actually matter.  If you translate the whole system in one direction, the symmetric copies get translated in a *different* direction.  So you can't just take whatever particle positions you're given, apply symmetry operators to them, and expect it to work.  Most likely you'll get horrible clashes with different copies right on top of each other.  You need to run an optimization over possible translations to find one that works as a starting conformation.

The other issue is that OpenMM expects the System and the Topology to have identical numbers of atoms.  With the API you suggested earlier, you would give it a Topology with N atoms, and then it would create a System with M*N atoms, where M is the number of symmetric copies per unit cell.  We need to figure out how to deal with this, which is partly a matter of figuring out how we want it to behave.  For example, if you write a PDB file, what should that file contain?  Just a single copy of the molecule?  Or all the different symmetric copies?

**Comment by jchodera** [comment]
> First of all, when you start adding symmetric copies, that suddenly means that global translations actually matter. If you translate the whole system in one direction, the symmetric copies get translated in a different direction. So you can't just take whatever particle positions you're given, apply symmetry operators to them, and expect it to work. Most likely you'll get horrible clashes with different copies right on top of each other. You need to run an optimization over possible translations to find one that works as a starting conformation.

If you start with a PDB structure, which already has the space group encoded in the file and coordinates that work within that space group.

If you start with a small molecule, yes, you have to worry about that, but you can get around that using an alchemically decoupled molecule that doesn't interact with itself.  Mike Schnieders has had great success using expanded ensemble / OSRW alchemical methods this way.

> The other issue is that OpenMM expects the System and the Topology to have identical numbers of atoms. With the API you suggested earlier, you would give it a Topology with N atoms, and then it would create a System with M*N atoms, where M is the number of symmetric copies per unit cell. We need to figure out how to deal with this, which is partly a matter of figuring out how we want it to behave. For example, if you write a PDB file, what should that file contain? Just a single copy of the molecule? Or all the different symmetric copies?

I see---this is a problem with information flow because the `System` has explicitly-represented virtual sites (which count as particles).  We have the same issue with virtual sites in the dummy atoms, and virtual sites in general, yes?  That means this should be in `Modeller` and not `ForceField`.

**Comment by jchodera** [comment]
> If you start with a PDB structure, this already has the space group encoded in the file and coordinates that work within that space group.

This argues that `PDBFile` and its `PDBML/mmCIF` friends should also store space group information that could potentially be used by `Modeller` if desired.

**Comment by peastman** [comment]
> If you start with a PDB structure, this already has the space group encoded in the file and coordinates that work within that space group.

No, we can't assume that.  Remember, we want to compute the positions of the symmetric atoms as

**x**' = **Rx** + u**a** + v**b** + w**c**

The PDB coordinates won't necessarily satisfy that.  The best we can say about them is

**x**' ≈ **R**(**x**-**y**) + u**a** + v**b** + w**c**

That is, there's some global translation **y** we can apply to the coordinates that will make them satisfy the first equation, but we don't know what it is.  Also, notice that I replaced = by ≈.  Physically, the copies are in symmetric positions.  But generally they'll have solved for the unit cell as a whole, so they won't necessarily be exact mathematical transformations of each other.

So here's what we get from loading a PDB file.  We *might* know the symmetry group, but very likely the file will have gotten processed and that information will have been lost along the way.  It will contain multiple chains which contain similar (but not necessarily identical) lists of atoms in positions that are roughly symmetric in a way that we haven't been told, but we can try to figure out by looking at them.

Of course, it's equally likely that the extra copies will already have been stripped out in an earlier processing step.

**Comment by jchodera** [comment]
Hm.  Try this: Load a PDB file into PyMOL and type
```
symexp sym,all,(all),5
```
Somehow, the PyMOL [`symexp`](http://pymolwiki.org/index.php/Symexp) keyword can easily figure this out from the information in the PDB file.

**Comment by jchodera** [comment]
That page also links to [this simple code](http://pymolwiki.org/index.php/Supercell) to fill in symmetry-related copies in the supercell.  It uses the `xray.sg_sym_to_mat_list(spacegroup)` function in PyMOL to get the rotation/translation matrices associated with each space group.

**Comment by jchodera** [comment]
Note I'm only talking about *actual* PDB files from the Protein Databank here. A non-compliant PDB file that does not represent a protein in a crystal doesn't necessarily satisfy that, but the tens of thousands of structures a user might download from the protein databank would.

**Comment by peastman** [comment]
It looks like they're just subtracting off the mean position, which is a reasonable rough guess for the needed translation.  I was figuring on doing that or something similar as the starting guess for an optimization.  It's probably good enough for visualization, since it's not important if a few residues end up on top of each other, but it's not nearly good enough for a simulation.

Also remember that people don't usually use files from RCSB directly.  They'll first have processed it in some other program (maybe PDBFixer, maybe something else), and most annotations will have been lost at that stage.

**Comment by jchodera** [comment]
> Also remember that people don't usually use files from RCSB directly. They'll first have processed it in some other program (maybe PDBFixer, maybe something else), and most annotations will have been lost at that stage.

We can do a few things to help this:
* Have `PDBFile` and friends store and expose symmetry information we can pass to `Modeller`
* Preserve symmetry information in reading/writing/processing when sensible to do so, invalidating it when we've done something to destroy it

**Comment by peastman** [comment]
True.  This also probably isn't a major concern.  If we require the user to tell us what symmetry group to use, that's not a serious burden on them.  I'm more concerned about the other issues I brought up above: figuring out where to position the system so symmetric copies don't clash, and deciding how to deal with systems and topologies.

Note this is very different from all the other types of virtual sites that ForceField normally deals with.  In those cases, a residue has some normal atoms, and matches a template based on them, and then computes a virtual site based on their positions.  In this case, we have entire chains that consist of nothing but virtual sites.  But they still need all the standard energy terms applied to them.

**Comment by jchodera** [comment]
What if the main chain has virtual sites?  Can we have virtual sites on virtual sites?

**Comment by peastman** [comment]
>  Can we have virtual sites on virtual sites?

Nice try, but no. :)

**Comment by peastman** [comment]
I do see your concern though.  Hmmm...

**Comment by peastman** [comment]
In this case we might be able to make an exception.  Generally speaking, the problem with one virtual site depending on another one is that all virtual sites get computed in parallel.  You can't have one position depend on other positions that may not have been computed yet.  But in this case, it only depends on one other particle.  So it wouldn't be difficult to have it just go ahead and redundantly compute that particle's position, then compute another one based on it.

**Comment by jchodera** [comment]
OK!  

What do you think the `Modeller`-based high-level API would look like?

**Comment by peastman** [comment]
I'm not really sure what the high level API would look like.  Let's continue trying to build it up from the lower level.

Suppose we allow you to define the symmetric virtual sites, as described above.  What then?  For example, suppose you have an ordinary system, and now you want it to create three extra symmetric copies and simulate that as a whole.  I *think* the following would work?

1. Add the extra particles to the system, setting them as massless and marking them as virtual sites.

2. Add the extra particles to any nonbonded forces, using identical parameters and exceptions.

3. Whenever you compute the force or energy, divide the part from nonbonded interactions by 4.

Alternatively you could *also* copy over all bonded terms so they would be applied to the symmetric copies.  Then you would have to divide the *total* force and energy by 4.

A different way of implementing this would be to do everything in the integrator.  There wouldn't be any virtual sites.  Everything would be real particles.  But when you integrate them, you average the forces from the symmetric copies of an atom, use that to update just one of them, and then set all the others to be symmetric versions of that.

**Comment by jchodera** [comment]
> A different way of implementing this would be to do everything in the integrator. 

Are you thinking of implementing this at the very basic `Integrator` level under the hood? If not, I think it would be very limiting to only have a specific kind of `SpaceGroupSymmetryLangevinIntegrator` or have to implement whatever dynamics you wanted in a `CustomIntegrator`. Let's explore the other options first.

As far as whether to only replicate the nonbonded forces or *all* forces, wouldn't it be simpler to simply replicate everything?

Would we actually need to divide forces by 4?  The real particles feel the correct forces from all the virtual sites. Only if the virtual sites also have forces computed *and* those forces are transferred to the real particles would we need to divide forces by 4. Instead, why not simply skip computing the forces on the virtual sites and not transfer those forces back to the original site. Only if we are computing the Hessian would those interactions matter anyway, right? Or does this break the notion of virtual sites, even if we need a new type of virtual site anyway?

I'll again go back to the high-level API because I think it's helpful to think of how people might actually *use* this feature to, say, simulate a crystal structure.  Here's an example:
```python
# PDBFile can't retrieve a PDB file from the RCSB, so use PDBFixer
pdbfixer = PDBFixer(pdbid='2RJY')
# Set up a Modeller instance, specifying the space group.
# Modeller would handle the replication for a given space group.
modeller = Modeller(pdbfixer.topology, pdbfixer.positions, spacegroup=pdbfixer.spacegroup)
# Alternatively, we could require a specific command, like:
modeller.addSpaceGroupSymmetry(spacegroup=pdbfixer.spacegroup)
# Create the system...
system = forcefield.createSystem(modeller.topology, nonbondedMethod=PME)
```
Somewhere in `Modeller`, we would create the virtual particles and the extra particle force terms automatically, so that the user doesn't have to worry about this.

**Comment by peastman** [comment]
> Or does this break the notion of virtual sites, even if we need a new type of virtual site anyway?

Well, it certainly conflicts with how all existing types of virtual sites work.  And I'm not convinced the particles really do feel the "correct" force.  The force needs to be the gradient of the energy.  If the position of particle B depends on the position of particle A, and if the energy depends on the position of particle B, that implies a contribution to the force on A.

**Comment by jchodera** [comment]
I think you're right.

Suppose we have U(x,y), but slave the position y to x via the function f(x).  So we now have U(x,f(x)).

Then:
```
\partial U/ \partial x = \partial U / \partial x + (\partial U / \partial y) (\partial f / \partial x)
```
so the force will depend on both via the chain rule.

**Comment by tristanic** [comment]
(Cross-posting from the OpenMM forum, and adding a few other thoughts)

I'm very interested in having this functionality available for [ISOLDE](https://www.researchgate.net/project/Next-generation-tools-for-building-atomic-models-into-challenging-experimental-data), and would be happy to have a go at tackling the VirtualSite implementation (starting in a few weeks or so, once my next version is ready to push out). Really handling it properly will involve dealing with various tricky corner cases, but I think there's a lot that can be handled incrementally.

First, though, a more general comment. I'm very much in agreement with the general thrust of the paper Peter linked in the OP, but an experimental crystallographer would argue that it's coming in after all the "hard work" has been done - that is, a "nearly correct" model has already been built into the experimental map, and the simulation is being used to do the final, very-fine-grained refinement. Particularly when pushing towards the lower resolutions (beyond about 2 Angstroms), I would very strongly advise that except in very rare circumstances this approach will just lead to garbage-in/garbage-out scenarios. Model building into low resolution maps is very hard with current tools, and the average worse-than-2-Angstrom model currently on the PDB can be expected to contain at least a few errors that no amount of "blind" MD will correct. This is inadvertently illustrated by this paper from the same group: https://www.sciencedirect.com/science/article/pii/S0006349515006761.  If you look at the tip of the hairpin illustrated in Figure 4, you'll see that yes: AMOEBA improved the fit to the map, but it didn't actually fix the underlying error. Look closely and you'll see two _cis_ peptide bonds, which I can tell you at a glance are indisputably wrong. In reality these (other than those for which the second residue is proline) occur at a rate of ~3/10,000 residues, and only in highly constrained/stabilised sites. Just one of many potential errors (of varying level of severity) that creep into models. That's why my focus is on bringing MD right back towards the _start_ of the model-building process, and combining it with continuous validation and an API to interactively (or, eventually, programmatically) find and correct problems in an efficient manner. In that context, it's rare to actually want to do a complete, PME simulation in P1 - the vast majority of the time is spent in small, localised "micro-simulations" to quickly fix specific issues, with full-scale simulations reserved for settling the final model at low temperature.

With that out of the way, a few of the more detailed issues that will come up:

- it's not all that uncommon for a deposited structure to have one or more residues with broken symmetry at the interface with symmetry copies. A classic example is where an asparagine or glutamine side chain H-bonds to its own rotationally symmetric equivalent. In that case one terminal amide has to be flipped 180 degrees relative to the other. Interestingly enough, the PDB architecture doesn't have the framework to annotate such cases, so you can't even rely on reading the metadata to find them. There are various other similar situations, where you'd either need to start implementing multiple non-interacting conformers of the offending residues, or doing a preliminary expansion to a lower space group, assignment of compatible rotamers, and then work up. If you want to work in explicit solvent, then the former approach is probably out of the question.

- It's possible to have bonds (most commonly disulfides) between symmetry neighbours. The challenges involved in handling this are fairly similar to the above.

- It's also (somewhat more) common to find symmetric small molecules/ions on special positions, where one or more symmetry operators maps the position back onto itself. Again, something that needs to be caught and specially handled. Current main-stream crystallographic refinement packages tend to handle this by recognising molecules "close to" special positions, moving them exactly *to* the special position, constraining them to a line (for a 2-fold axis) or a point (any higher symmetry), and excluding them from symmetry calculations.

- If you're going to use explicit solvent, it's _probably_ better to have some hybrid scheme where the macromolecules are symmetric but the solvent isn't. Otherwise, how do you handle the situation where a solvent molecule approaches its own symmetry copy/ies? They can't repel each other, since then you'd end up with an artificial solvent exclusion zone along all symmetry axes. But allowing them to pass through each other also raises issues with maintaining physical realism in the surroundings. One could consider some hybrid scheme where crystallographically-resolved solvent molecules are symmetric while the remainder are not.

- Also to do with explicit solvent is the challenge of implementing a suitable barostat when your box dimensions are no longer free to change. How would this be done? Perhaps by defining (a) site(s) far from the protein molecule where individual solvent molecules can be gradually faded in/out of existence to maintain the desired pressure?

All non-trivial problems, which is why (for now, at least), I've come to the conclusion that explicit solvent in this context is more trouble than it's worth.

**Comment by tristanic** [comment]
> Suppose we allow you to define the symmetric virtual sites, as described above. What then? For example, suppose you have an ordinary system, and now you want it to create three extra symmetric copies and simulate that as a whole. I think the following would work?
> 
>     Add the extra particles to the system, setting them as massless and marking them as virtual sites.
>     Add the extra particles to any nonbonded forces, using identical parameters and exceptions.
>     Whenever you compute the force or energy, divide the part from nonbonded interactions by 4.
> 
> Alternatively you could also copy over all bonded terms so they would be applied to the symmetric copies. Then you would have to divide the total force and energy by 4.

Wouldn't it be sufficient to exclude each virtual particle from non-bonded interactions with its own symmetry group?

**Comment by tristanic** [comment]
> Wouldn't it be sufficient to exclude each virtual particle from non-bonded interactions with its own symmetry group?

Actually, I hadn't thought that fully through. In a complete system (that is, a full asymmetric unit expanded out to P1 in a periodic simulation) a VirtualSymmetrySite shouldn't count contributions from _any_ other VirtualSymmetrySite - all the nonbonded force contributions should be covered by interactions with the real atoms. But if trying to simulate any arbitrary _piece_ of the whole that's no longer true, making it far harder to capture all the force terms without double-counting. So in that case it probably is best to go with Peter's suggestion of calculating all nonbonded interactions for all symmetry particles, and dividing by the number of copies.

**Comment by tristanic** [comment]
Sorry for the necropost, but while working on a little tool for visualisation of crystal symmetry axes, I came across this nice example of the sort of problem that will make space group symmetry-aware MD simulation of an arbitrary model really challenging to implement.

The model here is 6qiy and the residue pictured is Tyr44, which lies next to a 2-fold symmetry axis (the pink/white line running ~horizontally through the image).  It's therefore in contact with its own symmetry equivalent (the lighter coloured one above is from the "real" asymmetric unit, the darker one is the symmetry copy). You can see pretty clearly from the difference map (the green wireframe effectively showing where more atoms should be) that it's modelled wrongly - in fact, the PDB specification offers no way to unambiguously model this particular scenario correctly!

What's actually going on here is that the symmetry breaks at this site. The tyrosine has two alternate conformations: one pointing left, one pointing right, and at each instance of this interface the two chains have to choose which Tyr gets which conformation. This arrangement (where the 'A' altloc of one residue must contact the 'B' altloc of a symmetry-related copy) was never envisaged when the PDB specification was written, but it crops up with surprising regularity.

One way to get around this problem would of course be to expand out to a lower symmetry spacegroup where you now have two explicit copies of the chain, but if your goal is to accurately model the experimental data that approach isn't necessarily correct. There are two possible scenarios in situations like this: (a) there is a strict relationship throughout the crystal as to which chain gets which altloc, in which case expanding to the lower symmetry is correct (but generally considered problematic when the site in question is a tiny fraction of the model - you're effectively doubling the number of fitting parameters for the same amount of data); or (b) the choice at each interface is random, in which case expanding the symmetry is unambiguously the wrong thing to do since by doing so you're imposing an order that doesn't exist in reality.

In general, the best approach would probably be to split this residue (and probably the two flanking residues to ensure sufficient flexibility) into two non-interacting copies, and then ensure that these interact sensibly with their symmetry mates. 

This contrasts with the also-common situation where you have a small, symmetric (or nearly so) molecule sitting exactly _on_ a symmetry axis. Here again it probably makes most sense to have multiple non-interacting copies (essential when the molecule in question isn't perfectly symmetric), but now the copies **can't** interact with their symmetry mates.

I'm sure issues like this can be overcome, but I suspect there's a rather large and time-consuming research project involved in getting it all right.

![6qiy_symmetry_clash](https://user-images.githubusercontent.com/22666848/73751788-b1446d80-4757-11ea-9ea1-b6483ab7f7ea.jpg)

**Comment by peastman** [comment]
Thanks!  It's really interesting to know that cases like this are common.

**Comment by tristanic** [comment]
The subject of simulations in space group symmetry came up in a discussion with @jchodera over the weekend, and got me thinking about the challenge again - and what I think is the core stumbling block crystallised for me (pardon the pun). The problem is that macromolecular crystals aren't like small-molecule crystals in one really important respect: typically less than 50% (sometimes _much_ less) of their matter is actually crystalline. 

The image below is from a fairly extreme case. We're looking down one axis of 3io0, and the radius of the sphere of displayed density is 150 Angstroms. In this case, the protein only makes up 14.45% of the crystal volume - the rest is solvent. I think it makes the problem clear: the only atoms that "know" they're part of a crystal are those that are part of a (semi-)rigid lattice making contacts to more than one neighbouring asymmetric unit (ASU) - that is, (most of) the protein and the most tightly bound waters and ions. Outside of the first/second layer of bound solvent, there is absolutely no physical principle saying that a water in a given position in _one_ ASU should be in the same position in _every_ ASU. In fact, we can tell from the x-ray data that this isn't the case - despite many attempts by many groups to come up for a better model for the bulk solvent component of a crystal, so far nobody's ever come up with anything that performs better than the assumption that it averages out to perfectly flat (that is, perfectly disordered). To a lesser extent, this problem also extends to sidechains pointing into these void regions: while their range of motion may be constrained by the positions of their backbone atoms, they're otherwise entirely independent of their symmetry-related equivalents.

The upshot, I think, is that attempting to implement "pure" space-group symmetry explicit-solvent simulations where everything is contained within a single asymmetric unit will be highly problematic and prone to all sorts of unphysical artifacts (mostly boiling down to the same old "what do you do with a molecule that approaches its own symmetry copy/ies?"). Beyond that, by imposing a long-range order that doesn't actually exist, it'd be inconsistent with the experimental data.

That's not to say the idea of symmetry-enabled simulations doesn't have merit - it does, and done well it would be a huge boon to experimental model-building tools. But the challenges can't be ignored. The obvious initial solution is to implement it only for vacuum or implicit-solvent simulations (or implicit solvent plus structurally-important waters). But I wonder if there's the theoretical possibility to develop a hybrid scheme where individual atoms can be designated as either strictly symmetrical or not... the simulation would then consist of a P1 box filled with the atoms designated asymmetric; positions of symmetric atoms would be instances of a single entity, interacting with all atoms surrounding each instance point. The questions are: (a) would it make physical sense (I think it probably would); and (b) would it actually offer any performance improvement over just running the P1 simulation with strong symmetry restraints/constraints?
 
![3io0_150A_sphere](https://user-images.githubusercontent.com/22666848/91828817-857b5e80-ec38-11ea-9d58-c66f816c4c54.jpg)

**Comment by peastman** [comment]
Thanks for the discussion!  How different do you think this is from standard solutions of proteins in solvent, where we arbitrarily impose periodic boundary conditions?  In that case too we impose order on the simulation that doesn't exist in the physical system.  The only justification is that empirically it doesn't seem to cause any serious problems.  Are there reasons to expect it to cause more problems in crystal simulations?  I suppose one effect of imposing higher symmetries is that it leads to a smaller periodic unit, which could make artifacts more pronounced.

**Comment by tristanic** [comment]
Well... I suppose most of the effects of imposing strict symmetry throughout the asymmetric unit could be dealt with (e.g. by accumulating fluctuations in the time domain rather than across individual copies). But the big difference between this and standard periodic boxes is still that in the latter, a given atom never encounters its own periodic image(s). I think the challenge will really become apparent in a space group like P 6 2 2. The image below is from 6cy6, looking straight down the 6-fold axis. The "real" asymmetric unit is the brighter-coloured atoms at bottom; the rest are the symmetry copies. The question is: what happens to waters as they approach the 6-fold? If they're able to feel each other they'll get more and more crowded against their own symmetry equivalents, until at equilibrium I suppose you'd end up with a frozen hexagonal wall surrounding a tube of vacuum. I don't think there's a one-size-fits-all solution to this. You could implement a rule such as "a solvent molecule doesn't interact with its own symmetry equivalent" which would circumvent the "frozen wall" problem, but in some situations you do in fact see well resolved waters and other small molecules H-bonding to their symmetry equivalent(s). 

![6_fold_axis](https://user-images.githubusercontent.com/22666848/91904021-1df6fa80-ec9c-11ea-8f17-0862de5df11d.jpg)

**Comment by peastman** [comment]
Yes, I can see how that would cause problems.  You always want to keep lots of space between periodic copies of the same atom.

**Comment by aizvorski** [comment]
Only now found this thread, really nice - this is what I was trying to do in https://github.com/openmm/openmm/issues/3479  I'd also be interested in trying to get symmetrical simulations to work.

**Comment by aizvorski** [comment]
@peastman @tristanic  I wanted to ask about @jchodera's "Suppose we have U(x,y), but slave the position y to x via the function f(x). So we now have U(x,f(x))."   Is it fair to say this cannot be done with virtual sites?  

Virtual sites can interact with particles (eg using nonbonded force), and would effectively transmit forces to the particles that are used to define them, but do they interact also with the particles that are used to define them? (assuming both are within the nonbonded cutoff)  How about with all other virtual sites defined from the same particles?  And how is chaining of derivatives handled in those cases?

There's at least three strategies to set up a symmetrical simulation:
1. Virtual sites: needs at least a new type of virtual site with symmetry, and would have to figure out how a virtual site interacting with the parent particle and all the other virtual sites which are symmetrical copies of the same particle, or
2. Real particles with a force that tries to bring them to symmetrical positions - like a harmonic restraint, but with each particle restrained to the average position of all copies, or
3. Real particles with symmetry constraints, similar to distance constraints

I'm not sure how self-interaction between copies would be handled in 1, and whether forces get transmitted from the virtual site to the parent correctly. That also probably needs a new type of VirtualSite subclass with a rotation+translation matrix since (I think?) an OutOfPlaneSite can't do this, even with extra particles to define the rotation axis.  The setup in 2 could be  implemented using CustomBondForce between all symmetrical copies, it probably doesn't require any new features not already in openmm, but doesn't enforce exact symmetry (also may need very strong force constant and/or short timesteps).  Something like 2 with weak force may be interesting for some of the partial-symmetry cases @tristanic mentioned.  Doing 3 requires a new kind of constraint, and would be handled during the constraints step of the integrator - but also this should be pretty easy to satisfy because if the positions are symmetrical, the forces are also symmetrical up to numerical accuracy.

What do you think about the pros and cons of each of these?

**Comment by tristanic** [comment]
> you generally add exceptions to keep them from interacting with the atoms they're computed from. But you don't need to.

But I think you _would_ need to prevent them from interacting with other virtual particles in the same symmetry group (because those interactions are already covered by the "real" atoms).

**Comment by peastman** [comment]
Or alternatively you could multiply the mass of each atom by the number of symmetric copies.  In that case you would also need to multiply the strength of the bonded terms by the same factor.

**Comment by aizvorski** [comment]
> Or alternatively you could multiply the mass of each atom by the number of symmetric copies. In that case you would also need to multiply the strength of the bonded terms by the same factor.

@peastman Ah, nice!  I was assuming the mass would be multiplied but hadn't quite realized the bonded terms should be multiplied also.  I *think* that solves that problem nicely, for fully symmetrical systems.

For symmetrical/asymmetrical mix, suppose we have symmetrical atom a1 with virtual sites va2, va3 and this is bonded to asymmetrical atoms b1, b2, b3.  If the mass of a1 is 3x, but the mass of b1, b2, b3 is normal and the va2-b2, va3-b3 bonds have the normal strength, I think that also works correctly.

**Comment by aizvorski** [comment]
> imposing the symmetry introduces a _de facto_ friction term proportional to the number of symmetry copies

> once you get up to 5-fold symmetry or so, tugging on one atom is like trying to pull it through molasses

@tristanic That's a fascinating observation!  (fairly curious about what the interactive-MD setup looks like)  I think these approaches shouldn't in theory have any added friction from the _interaction_ between copies, if the external force is applied to all copies of a symmetrical atom simultaneously (transformed through the symmetry group of course).  But I can imagine how an external force applied to just a single copy would do that, yes.

**Comment by tristanic** [comment]
Yes - if the force is split between all copies of the atom then that would do the trick. Probably what I should have done back then - but NAMD's interactive MD interface was extremely​ limited, with all external forces calculated in Tcl in VMD (already way beyond what that language was designed for). Also, I was still very green.

Might also need to consider that when thinking about interactions with the thermostat - if it treats each symmetry copy as independent then the effective temperature will be much lower than expected.

As to what it look(s/ed) like: there's a video of my old VMD/NAMD implementation at https://www.youtube.com/watch?v=NyvVzn1kC-U. That was based around VMD's support for the old Novint Falcon haptic interface. Was cool, but extremely clunky in many ways. These days it's all about ISOLDE (https://isolde.cimr.cam.ac.uk) - see https://twitter.com/CrollTristan/status/1466829495558680579?s=20&t=TXrDpnHWQPOiPfC_z_mt9Q for an intro to the latest version.


________________________________

**Comment by peastman** [comment]
> Might also need to consider that when thinking about interactions with the thermostat - if it treats each symmetry copy as independent then the effective temperature will be much lower than expected.

All the symmetric copies are massless, so the integrator will ignore them.

**Comment by tristanic** [comment]
> All the symmetric copies are massless, so the integrator will ignore them.

That would be true for virtual sites, but not, I think, for options 2 and 3 in @aizvorski's list.

**Comment by peastman** [comment]
In option 2 they're ordinary particles moving in a potential, so the thermostat should treat them like any other particles.

**Comment by tristanic** [comment]
Potentially of interest: https://twitter.com/andrewwhite01/status/1501375768089313282?s=20&t=eMTamyTUWE2PbMywivfTrw

**Comment by whitead** [comment]
Wow I wish I had noticed this issue before beginning our project! Thanks @tristanic for messaging me about it. Lots of interesting discussion. 

We implemented this idea, which we called symmetric MD, in our own [simple engine to test out the ideas](https://github.com/whitead/symd). Here is a short [thread on Twitter](https://twitter.com/andrewwhite01/status/1511138339998650378) with some animations/explanations.

We found that indeed integrating the asymmetric unit only and making the other particles virtual will give the correct answer, with some care in counting degrees of freedom. We also did the difficult task of making all the affine transformations, Bravais lattices, and asymmetric unit definitions accessible in a python package on pypi called `symd`

I'll paste here my notes on how I think it could be done in modern simulation engines. Some of these notes are not needed if you're restricting yourself to NVT. 

> Here is our advice on implementing symmetric molecular dynamics in a modern molecular dynamics engine. The asymmetric unit should be integrated as usual. Make the non-asymmetric unit particles (images) be ghost particles;  ghost particles are non-integrated particles used in force-field calculations. The ghost particles' positions should be set using affine matrices defining the group transformations in fractional coordinates. These matrices can be obtained from our library or crystallography tables.

>Pressure computed for the asymmetric unit is not meaningful and NPT should be done using the algorithm described above that does not require computing pressure. The lattice vectors may be stored separately than the usual lattice vectors and are only used to set the ghost particle positions.  The tensor transforms can be loaded from our library. Periodic boundary conditions should be disabled entirely if doing NPT.

> The constraints for Wyckoff sites are implemented as Lagrange multiplier constraints. The terms can be computed analytically at each step using Equation 10

See the [complete preprint](https://arxiv.org/abs/2204.01114)

**Comment by peastman** [comment]
This is implemented in #4185.
